indico/modules/events/management/controllers.py

```diff
@@ -328,6 +400,89 @@ def _process(self):
         return jsonify_data(flash=False)
 
 
+class RHClonePreview(RHManageEventBase):
+    def _process(self):
+        form = CloneRepeatabilityForm()
+
+        if form.repeatability.data == 'interval':
+            clone_calculator = IntervalCloneCalculator(self.event_new)
+        elif form.repeatability.data == 'pattern':
+            clone_calculator = PatternCloneCalculator(self.event_new)
+        else:
+            raise BadRequest
+
+        try:
+            dates = clone_calculator.calculate(request.form)
+            if len(dates) > 100:
+                raise ValueError(_("You can clone maximum of 100 times at once"))
+        except ValueError as e:
+            return jsonify(error={'message': e.message})
+        return jsonify_data(count=len(dates), dates=dates)
+
+
+class RHCloneEvent(RHManageEventBase):
+    """Create copies of the event."""
+
+    def _form_for_step(self, step, set_defaults=True):
+        if step == 1:
+            return CloneRepeatabilityForm()
+        elif step == 2:
+            return CloneContentsForm(self.event_new, set_defaults=set_defaults)
+        elif step == 3:
+            default_category = (self.event_new.category if self.event_new.category.can_create_events(session.user)
+                                else None)
+            return CloneCategorySelectForm(self.event_new, category=default_category)
+        elif step == 4:
+            return REPEAT_FORM_MAP[request.form['repeatability']](self.event_new, set_defaults=set_defaults)
+        else:
+            return None
+
+    def _process(self):
+        step = int(request.form.get('step', 1))
+        tpl_args = {}
+        form = self._form_for_step(step, set_defaults=True)
+        prev_form = self._form_for_step(step - 1)
+
+        if prev_form and not prev_form.validate():
+            form = prev_form
+            step = step - 1
+
+        if step == 4:
+            tpl_args.update({
+                'step_title': dict(CLONE_REPEAT_CHOICES)[request.form['repeatability']],
+            })
+        elif step > 4:
+            # last step - perform actual cloning
+            form = REPEAT_FORM_MAP[request.form['repeatability']](self.event_new)
+
+            if form.validate_on_submit():
+                if form.repeatability.data == 'once':
+                    dates = [form.start_dt.data]
+                else:
+                    if form.repeatability.data == 'interval':
+                        clone_calculator = IntervalCloneCalculator(self.event_new)
+                    elif form.repeatability.data == 'pattern':
+                        clone_calculator = PatternCloneCalculator(self.event_new)
+                    else:
+                        raise BadRequest
```

Comment:
The above if-else block could go into a function since it's repeated in `RHClonePreview`